pkg/server/testing/fake_images_client.go

```diff
@@ -0,0 +1,168 @@
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/containerd/containerd/api/services/images"
+	googleprotobuf "github.com/golang/protobuf/ptypes/empty"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// FakeImagesClient is a simple fake images client, so that cri-containerd
+// can be run for testing without requiring a real containerd setup.
+type FakeImagesClient struct {
+	sync.Mutex
+	called    []CalledDetail
+	errors    map[string]error
+	ImageList map[string]images.Image
+}
+
+var _ images.ImagesClient = &FakeImagesClient{}
+
+// NewFakeImagesClient creates a FakeImagesClient
+func NewFakeImagesClient() *FakeImagesClient {
+	return &FakeImagesClient{
+		errors:    make(map[string]error),
+		ImageList: make(map[string]images.Image),
+	}
+}
+
+// getError get error for call
+func (f *FakeImagesClient) getError(op string) error {
+	err, ok := f.errors[op]
+	if ok {
+		delete(f.errors, op)
+		return err
+	}
+	return nil
+}
+
+// InjectError inject error for call
+func (f *FakeImagesClient) InjectError(fn string, err error) {
+	f.Lock()
+	defer f.Unlock()
+	f.errors[fn] = err
+}
+
+// InjectErrors inject errors for calls
+func (f *FakeImagesClient) InjectErrors(errs map[string]error) {
+	f.Lock()
+	defer f.Unlock()
+	for fn, err := range errs {
+		f.errors[fn] = err
+	}
+}
+
+// ClearErrors clear errors for call
+func (f *FakeImagesClient) ClearErrors() {
+	f.Lock()
+	defer f.Unlock()
+	f.errors = make(map[string]error)
+}
+
+func (f *FakeImagesClient) appendCalled(name string, argument interface{}) {
+	call := CalledDetail{Name: name, Argument: argument}
+	f.called = append(f.called, call)
+}
+
+// GetCalledNames get names of call
+func (f *FakeImagesClient) GetCalledNames() []string {
+	f.Lock()
+	defer f.Unlock()
+	names := []string{}
+	for _, detail := range f.called {
+		names = append(names, detail.Name)
+	}
+	return names
+}
+
+// SetFakeImages injects fake images.
+func (f *FakeImagesClient) SetFakeImages(images []images.Image) {
+	f.Lock()
+	defer f.Unlock()
+	for _, image := range images {
+		f.ImageList[image.Name] = image
+	}
+}
+
+// Get is a test implementation of images.Get
+func (f *FakeImagesClient) Get(ctx context.Context, getOpts *images.GetRequest, opts ...grpc.CallOption) (*images.GetResponse, error) {
+	f.Lock()
+	defer f.Unlock()
+	f.appendCalled("get", getOpts)
+	if err := f.getError("get"); err != nil {
+		return nil, err
+	}
+	image, ok := f.ImageList[getOpts.Name]
+	if !ok {
+		return nil, fmt.Errorf("image does not exist")
+	}
+	return &images.GetResponse{
+		Image: &image,
+	}, nil
+}
+
+// Put is a test implementation of images.Put
+func (f *FakeImagesClient) Put(ctx context.Context, putOpts *images.PutRequest, opts ...grpc.CallOption) (*googleprotobuf.Empty, error) {
+	f.Lock()
+	defer f.Unlock()
+	f.appendCalled("put", putOpts)
+	if err := f.getError("put"); err != nil {
+		return nil, err
+	}
+	_, ok := f.ImageList[putOpts.Image.Name]
```

Comment:
I don't think image service will return error if the key already exists.
They are using boltdb https://github.com/containerd/containerd/blob/master/images/storage.go#L100, which doesn't return error if key already exists https://github.com/boltdb/bolt/blob/2c3edd782c07f917541f3094a42149bd7e497dc2/bucket.go#L282.